Source/Diagnostics/FieldIO.cpp

```diff
@@ -1,9 +1,193 @@
+            {openPMD::UnitDimension::I,  1},
+        });
+    } else if (field_name.substr(0,3) == "rho"){ // charge density
+        mesh.setUnitDimension({
+            {openPMD::UnitDimension::L, -3},
+            {openPMD::UnitDimension::I,  1},
+            {openPMD::UnitDimension::T,  1},
+        });
+    }
+}
+
+
+/** \brief
+ * Convert an IntVect to a std::vector<unsigned long long>,
+ * and reverse the order of the elements
+ * (used for compatibility with the openPMD API)
+ */
+std::vector<unsigned long long>
+getReversedVec( const IntVect& v )
+{
+  // Convert the IntVect v to and std::vector u
+  std::vector<unsigned long long> u = {
+    AMREX_D_DECL(
+                 static_cast<unsigned long long>(v[0]),
+                 static_cast<unsigned long long>(v[1]),
+                 static_cast<unsigned long long>(v[2])
+                 )
+  };
+  // Reverse the order of elements, if v corresponds to the indices of a
+  // Fortran-order array (like an AMReX FArrayBox)
+  // but u is intended to be used with a C-order API (like openPMD)
+  std::reverse( u.begin(), u.end() );
+
+  return u;
+}
+
+/** \brief
+ * Convert Real* pointer to a std::vector<double>,
+ * and reverse the order of the elements
+ * (used for compatibility with the openPMD API)
+ */
+std::vector<double>
+getReversedVec( const Real* v )
+{
+  // Convert Real* v to and std::vector u
+  std::vector<double> u = {
+    AMREX_D_DECL(
+                 static_cast<double>(v[0]),
+                 static_cast<double>(v[1]),
+                 static_cast<double>(v[2])
+                 )
+  };
+  // Reverse the order of elements, if v corresponds to the indices of a
+  // Fortran-order array (like an AMReX FArrayBox)
+  // but u is intended to be used with a C-order API (like openPMD)
+  std::reverse( u.begin(), u.end() );
+
+  return u;
+}
+
+/** \brief Write the `ncomp` components of `mf` (with names `varnames`)
+ * into a file `filename` in openPMD format.
+ **/
+void
+WriteOpenPMDFields( const std::string& filename,
+                  const std::vector<std::string>& varnames,
+                  const MultiFab& mf, const Geometry& geom,
+                  const int iteration, const double time )
+{
+  BL_PROFILE("WriteOpenPMDFields()");
+
+  const int ncomp = mf.nComp();
+
+  // Create a few vectors that store info on the global domain
+  // Swap the indices for each of them, since AMReX data is Fortran order
+  // and since the openPMD API assumes contiguous C order
+  // - Size of the box, in integer number of cells
+  const Box& global_box = geom.Domain();
+  auto global_size = getReversedVec(global_box.size());
+  // - Grid spacing
+  std::vector<double> grid_spacing = getReversedVec(geom.CellSize());
+  // - Global offset
+  std::vector<double> global_offset = getReversedVec(geom.ProbLo());
+  // - AxisLabels
+#if AMREX_SPACEDIM==3
+  std::vector<std::string> axis_labels{"x", "y", "z"};
+#else
+  std::vector<std::string> axis_labels{"x", "z"};
```

Comment:
wow, not x and y?

Reply:
Yep, Warp convention. 😄

Reply:
Cannot complain, we propagate along `y` :smile: